### PR TITLE
CFY-5205 Add deployment update documentation

### PR DESCRIPTION
**Pull request by AviaE**

### CODE REVIEW

File: content/manager/update-deployment.md [hunk 1]
```diff
@@ -6,58 +6,70 @@ draft: false
...
+The contents of the deployment update should be described in a [yaml blueprint file](http://docs.getcloudify.org/3.4.0/blueprints/overview/), just as any application in Cloudify. Following the aforementioned example, The updated application blueprint will probably include a new database type, a few new node templates of the new database type, and a few new relationships representing how these new nodes should be connected to the existing architecture.
 
 ### Using the CLI to Update a Deployment
 One quick way to update your deployment with Cloudify is using the CLI. Another way, perhaps more 'visual' is using the Cloudify UI. Updating a deployment via the CLI is quite reminiscent of uploading a blueprint or creating a deployment. You'll need a blueprint file describing your deployment update. That blueprint can be uploaded directly by supplying a local file path, or it can be upload as an archive.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: can be uploaded

File: content/manager/update-deployment.md [hunk 2]
```diff
@@ -6,58 +6,70 @@ draft: false
...
-##### via an archived blueprint
+#### via an archived blueprint
 ```shell
 cfy deployments update -d ID_OF_DEPLOYMENT_TO_UPDATE -l ARCHIVE_PATH [-n BLUEPRINT_FILENAME]
 ```
-When updating a deployment using an archive, the name of the blueprint representing the deployment update is assumed to be `blueprint.yaml`. If the blueprint file has a different name, it must be specified using the `-n / --blueprint-filename` optional argument.
+When updating a deployment using an archive, the name of the blueprint representing the deployment update is assumed to be `blueprint.yaml`. If the blueprint file has a different name, it must be specified using the `-n / --blueprint-filename` argument.
 
-{{% gsInfo title="Some Deployment Update Concepts" %}}
-* The **deployment update blueprint** is the blueprint that contains the changes representing the deployment update, what was previously referenced as the 'updated application blueprint'.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: why use `updated application blueprint` at all? just use `deployment update blueprint`

File: content/manager/update-deployment.md [hunk 3]
```diff
@@ -6,58 +6,70 @@ draft: false
...
+{{% gsInfo title="Some deployment update concepts" %}}
+- The **deployment update blueprint** is the blueprint that contains the changes representing the deployment update, what was previously referenced as the 'updated application blueprint'.
+- A **step** is a logical concept that represents a single change in a deployment update blueprint. There are three different types of steps: **add**, **remove**, and **modify**. These three concepts will be used extensively throughout this guide. The scope of a 'step' is determined by its most top-level change. For example, if a node was added, and this node also contains a new relationship, then this is a 'add node' step, and not a 'add relationship' step. Similarly, if a node's property was modified, it is not a 'modify node' step, but a 'modify property' step. A list of all the possible steps is located [here](http://docs.getcloudify.org/3.4.0/manager/update-deployment/#what-can-be-updated-as-a-part-of-a-deployment-update)
 {{% /gsInfo %}}
+{{% gsNote title="Accessing the steps of a deployment update" %}}
+Currently, After you apply a deployment update, its composing steps are only accessible using the Cloudify REST API.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: What does it mean? what can I access? How?

File: content/manager/update-deployment.md [hunk 4]
```diff
@@ -6,58 +6,70 @@ draft: false
...
+{{% /gsNote %}}
 
 ### Deployment Update Flow
 Updating a deployment consists of several stages:
 1. The deployment update blueprint is uploaded to the manager.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: show each step in a different line.

File: content/manager/update-deployment.md [hunk 5]
```diff
@@ -6,58 +6,70 @@ draft: false
...
 2. The steps composing the deployment update are extracted.
 3. All the 'added' changes are updated in the data model.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: Namedropping of 'data model'. This term is not familiar to all. maybe add a link to an explanation, or don't mention.
Do most users even care about 'behind the scenes'? Maybe just add stuff like that as an advanced side note.

File: content/manager/update-deployment.md [hunk 6]
```diff
@@ -6,58 +6,70 @@ draft: false
...
 4. The `update` workflow is executed. As a part of the *default* update workflow execution:
-    * The `unlink` operation will be executed in regard the each removed relationship.
-    * The `uninstall` workflow will be executed on each of the removed nodes.
-    * The `install` workflow will be executed on each of the added nodes.
-    * The `establish` operation will be executed in regard the each added relationship.
+    - The `unlink` operation will be executed in regard the each removed relationship.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: The indentation is not presented in the site.
"in regard to each.."

File: content/manager/update-deployment.md [hunk 7]
```diff
@@ -6,58 +6,70 @@ draft: false
...
+    - The `uninstall` workflow will be executed on each of the removed nodes.
+    - The `install` workflow will be executed on each of the added nodes.
+    - The `establish` operation will be executed in regard the each added relationship.
 5. All the 'removed' changes are updated in the data model
-{{% gsNote title="Workflow/operation Execution during a deployment update" %}}
-Phase 4 of the deployment update flow includes the only cases in which a workflow or an operation is executed during a deployment update
+{{% gsNote title="Workflow/operation execution during a deployment update" %}}
+Phase 4 of the deployment update flow includes the only cases in which a workflow or an operation is executed during a deployment update. i.e. adding an operation, removing a workflow, modifying the `install-agent` property or any other step that is not add/remove node or relationship will not trigger an execution.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: Inconsistent. Switched between stages to phases.

File: content/manager/update-deployment.md [hunk 8]
```diff
@@ -6,58 +6,70 @@ draft: false
...
+{{% /gsNote %}}
+{{% gsNote title="make sure the built in update workflow is contained in your blueprint" %}}
```
Comment (nirbir) [inline content/manager/update-deployment.md]: built-in

File: content/manager/update-deployment.md [hunk 9]
```diff
@@ -6,58 +6,70 @@ draft: false
...
+With Cloudify, you can update a deployment that was previously [created from a blueprint](http://docs.getcloudify.org/3.4.0/manager/create-deployment/). But what does 'updating' a deployment mean? Well, say you have a sizable intricate deployment of webservers and databases. After some time and research, you realize that you need to add a new kind of database, that should be connected to some of the existing webservers. 'Updating' a deployment means that instead of *creating a new deployment* from a blueprint that includes these new servers, you can simply add and connect these new databases to your *existing deployment*, while retaining the state of your current webservers-databases setting.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: The side bar on the right doesn't show any of the headlines. Change the sizes.

File: content/manager/update-deployment.md [hunk 10]
```diff
@@ -6,58 +6,70 @@ draft: false
...
 ```
 If you choose to skip the `install` workflow, added nodes won't be installed, and added relationships won't be established
 ```shell
 cfy deployments update -d ID_OF_DEPLOYMENT_TO_UPDATE -p PATH_TO_BLUEPRINT --skip-uninstall
 ```
-If you choose to skip the `uninstall` workflow, removed nodes won't be uninstalled, and removed relationship won't be unlinked
+If you choose to skip the `uninstall` workflow, removed nodes won't be uninstalled, and removed relationship won't be unlinked.
+
+#### Recovering from a failed update
+If the deployment update workflow fails during its execution, you try to preform another deployment update to recover from it, this time using the `-f` flag. A common solution is to try to do a 'rollback', using a deployment update blueprint that represents the previous deployment.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: you can try

File: content/manager/update-deployment.md [hunk 11]
```diff
@@ -68,34 +80,32 @@ webserver:
     properties:
         port: {get_input: port}
 ```
-Now, suppose that while updating this deployment you overrode the `port` input with `9090`, and assume that the `webserver` node didn't change as part of the update. Which means, relying on the deployment update flow, that no install and/or uninstall workflows were run on this node. As a result, its `port` property is still `8080`. In contrast, any new nodes (including new `webserver` nodes) that were added as a part of that deployment update and use the `port` input, will be assigned with the new `port` input value - `9090`.
+Now, suppose that while updating this deployment you overrode the `port` input with `9090`, and assume that the `webserver` node didn't change as part of the update. Which means, relying on the deployment update flow, that no install and/or uninstall workflows were run on this node. As a result, its `port` property of that node is still `8080`. In contrast, any new nodes (including new `webserver` nodes) that were added as a part of that deployment update and use the `port` input, will be assigned with the new `port` input value - `9090`.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: 'you override the..'

File: content/manager/update-deployment.md [hunk 12]
```diff
@@ -68,34 +80,32 @@ webserver:
...
 {{% /gsNote %}}
 {{% gsNote title="Overriding default input values" %}}
-Similar to overriding existing inputs, changing the default values of inputs won't affect nodes that were already installed,
+Similar to overriding existing inputs, changing the default values of inputs won't affect nodes that were already installed.
 {{% /gsNote %}}
 
 ### Referencing Existing Resources and Uploading New Ones:
```
Comment (nirbir) [inline content/manager/update-deployment.md]: no need for ':'

File: content/manager/update-deployment.md [hunk 13]
```diff
@@ -68,34 +80,32 @@ webserver:
...
-{{% gsNote title="previously imported blueprints in the `inputs` section" %}}
-Unlike resources, entries from the [`imports`](http://docs.getcloudify.org/3.4.0/blueprints/spec-imports/) section that were part of that deployment's blueprint or of a previous deployment update must be imported in the deployment update blueprint as well. e.g if the blueprint of the original deployment contained within its imports the entry `http://www.getcloudify.org/spec/cloudify/3.4/types.yaml`, the same entry must be also under the `imports` section of the blueprint representing any of its deployment updates.
+Any previously uploaded resource (scripts, data files, etc.) can be referenced inside the deployment update blueprint. However, and this applies both to updating via an archive and via a blueprint file, uploading a resource as part of the update with the same name as an existing one will overwrite that resource throughout that deployment.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: both to -> to both

File: content/manager/update-deployment.md [hunk 14]
```diff
@@ -68,34 +80,32 @@ webserver:
...
 {{% /gsNote %}}
 
 ### Unsupported Changes in a Deployment Update
-If a deployment update blueprint contains changes that are not currently supported as a part of an update, the update will not take place, and a message indicating the unsupported changes will be displayed to the user. Following is a list of unsupported changes, along some possible examples of these changes.
-#### Node Type
+If a deployment update blueprint contains changes that are not currently supported as a part of an update, the update will not take place, and a message indicating the unsupported changes will be displayed to the user. Following is a list of unsupported changes, along some possible examples.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: along with

File: content/manager/update-deployment.md [hunk 15]
```diff
@@ -138,17 +146,16 @@ node_templates:
             properties:
                 connection_type: all_to_one  # unsupported update - can't modify a relationship's property
 ```
-#### operations implemented with plugins
+#### Operations implemented with plugins
 You cannot update a operation implemented with a plugin in the following cases:
```
Comment (nirbir) [inline content/manager/update-deployment.md]: update an...

File: content/manager/update-deployment.md [hunk 16]
```diff
@@ -138,17 +146,16 @@ node_templates:
...
-* The updated operation is implemented with a plugin that didn't exist in the original deployment
+- The updated operation is implemented with a plugin that didn't exist in the original deployment
```
Comment (nirbir) [inline content/manager/update-deployment.md]: indentation not presented

File: content/manager/update-deployment.md [hunk 17]
```diff
@@ -190,25 +194,22 @@ nodes:
...
 plugins:
     plugin1:
-        [...]
         install: true  # unsupported update - in the original deployment `plugin1` was different (its `install` was false)
 ```
-#### workflows plugin mappings
+#### Workflows plugin mappings
 You cannot update a workflow plugin mapping in the following case:
-* The plugin of the updated workflow, (whether the workflow currently exists or it is been added with the update) is not one of the current deployment plugins, and the `install` field of the updated workflow's plugin is `true`
+- The plugin of the updated workflow, (whether ifthe workflow currently exists or if it is been added with the update) is not one of the current deployment plugins, and the `install` field of the updated workflow's plugin is `true`.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: indentation

Comment (nirbir) [inline content/manager/update-deployment.md]: whether the workflow..

File: content/manager/update-deployment.md [hunk 18]
```diff
@@ -219,21 +220,19 @@ workflows:
...
-        [...]
         install: true
 ```
-#### Groups, Policy Types and Policy Triggers
-Any Change in the top level fields `groups`, `policy_types` and `policy_triggers` is not currently supported as a part of a deployment update blueprint
+#### Groups, policy types and policy triggers
+Any Change in the top level fields `groups`, `policy_types` and `policy_triggers` is not currently supported as a part of a deployment update blueprint.
 
 ### What Can be Updated as a Part of a Deployment Update
 The following can be updated as part of a deployment update, subject to the limitations that were [mentioned above](http://docs.getcloudify.org/3.4.0/manager/update-deployment/#unsupported-changes-in-a-deployment-update).
```
Comment (nirbir) [inline content/manager/update-deployment.md]: Hard coded 3.4.0? I believe there's a better way to add a link to a headline in the current page

File: content/manager/update-deployment.md [hunk 19]
```diff
@@ -219,21 +220,19 @@ workflows:
...
+Nodes can be added or removed, including all their relationships, operations, an so on. Remember that adding or removing a node will trigger the install/uninstall workflow in regard to that node.
+{{% gsNote title="'Renaming' nodes" %}}
+Assume that the original deployment blueprint contains a node named `node1`. Then, in the deployment update blueprint, you decide to 'rename' that node, to `node2`. Now the deployment update blueprint's `node2` is identical to `node1` in the original blueprint, except its name. But in practice, there isn't really a 'renaming' process. In the aforementioned scenario, `node1` will be uninstalled, and `node2` will be installed. that is `node1` won't retain its state.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: except for its name...

File: content/manager/update-deployment.md [hunk 20]
```diff
@@ -249,8 +248,8 @@ node_templates:
 ```
 {{% /gsNote %}}
 
-#### relationships
-except for being added or removed as part of adding or removing a node, relationships can be also be added or removed specifically. Adding a relationship will trigger execution of its `establish` operations (assuming a default `install` workflow). Similarly, removing an operation will trigger execution of the `unlink` operations. In addition, it is also possible to change a node's relationship order. The operations of the added and removed relationships will be executed according the order of the relationships
```
Comment (nirbir) [inline content/manager/update-deployment.md]: '... removed specifically' sounds weird.. maybe independently? we need another word here

File: content/manager/update-deployment.md [hunk 21]
```diff
@@ -249,8 +248,8 @@ node_templates:
...
+#### Relationships
+Except for being added or removed as part of adding or removing a node, relationships can be also be added or removed specifically. Adding a relationship will trigger execution of its `establish` operations (assuming a default `install` workflow). Similarly, removing an operation will trigger execution of the `unlink` operations. In addition, it is also possible to change a node's relationship order. The operations of the added and removed relationships will be executed according the order of the relationships in the deployment update blueprint.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: 'removing an operation' -> 'removing a relationship'

File: content/manager/update-deployment.md [hunk 22]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+            prop1: value1
+```
 ```yaml
 # deployment update blueprint
 node_templates:
-    node2:  # node1 will be uninstalled. node2 will be installed
+    node1:
+        type: Cloudify.nodes.Compute
+        properties:
+            ip: 192.0.2.1  # modified the property by overriding it default
```
Comment (nirbir) [inline content/manager/update-deployment.md]: What did we override? there was no default

File: content/manager/update-deployment.md [hunk 23]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+outputs:
+    workflow1:
+        value: plugin_name.module_name.task2  # modified the value of workflow1
+    # removed workflow2
+    workflow3:  # added workflow3
+        [...]
+```
+
+#### Description:
+The description can be added, removed or modified
```
Comment (nirbir) [inline content/manager/update-deployment.md]: seperate the 3 different examples. Maybe add mini-headlines - remove, add, modify

File: content/manager/update-deployment.md [hunk 24]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+```yaml
+# deployment update blueprint
+description: new_description
+```
 
+### Using a Custom Update Workflow
+If the default `update` workflow doesn't satisfy your every need, Cloudify enables you to create a custom `update` workflow to be used as part of the deployment update process.
+#### Requirements from the custom update workflow
+In order for the custom workflow to be compatible with the deployment update process, it must accept at least the following arguments:
+ * `ctx` - the regular ctx passed to any execution.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: indentation

File: content/manager/update-deployment.md [hunk 25]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+description: old_description
+```
+```yaml
+# deployment update blueprint
+description: new_description
+```
 
+### Using a Custom Update Workflow
+If the default `update` workflow doesn't satisfy your every need, Cloudify enables you to create a custom `update` workflow to be used as part of the deployment update process.
+#### Requirements from the custom update workflow
```
Comment (nirbir) [inline content/manager/update-deployment.md]: custom update workflow requirements

File: content/manager/update-deployment.md [hunk 26]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+In order for the custom workflow to be compatible with the deployment update process, it must accept at least the following arguments:
```
Comment (nirbir) [inline content/manager/update-deployment.md]: 'must accept the following...'

File: content/manager/update-deployment.md [hunk 27]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+ * `ctx` - the regular ctx passed to any execution.
+ * `update_id` - the id of the deployment update.
+ * `added_instance_ids` - the list of all the added node instances ids.
+ * `added_target_instances_ids` - the list of all the node instances ids which the added nodes have relationships with.
+ * removed_instance_ids - the list of all the removed node instances id.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: continue with the `` notation for the arguments names

File: content/manager/update-deployment.md [hunk 28]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+ * remove_target_instance_ids - the list of all the node instances id which the removed nodes had relationships with.
```
Comment (nirbir) [inline content/manager/update-deployment.md]: should this include a target instance which appeared in a removed node but still exists in a non removed node?

File: content/manager/update-deployment.md [hunk 29]
```diff
@@ -267,30 +266,227 @@ node_templates:
...
+      extended_instance_ids:
+        default: []
+      extend_target_instance_ids:
+        default: []
+      reduced_instance_ids:
+        default: []
+      reduce_target_instance_ids:
+        default: []
+```
+Furthermore, in order to finalize the deployment update (phase 5 of the [deployment udpate flow](http://docs.getcloudify.org/3.4.0/manager/update-deployment/#deployment-update-flow)), your custom `update` workflow must make a REST call, in the following manner:
```
Comment (nirbir) [inline content/manager/update-deployment.md]: phase 5 or stage 5?

File: content/manager/update-deployment.md (hunk 9 above)
Comment (AviaE) [inline content/manager/update-deployment.md]: Currently, the sidebar only shows `h1` and `h2` headers. But in my opinion, the `h2` font size is way too big. So I'm only willing to change my `h3`s to `h2` when the font sizes will be revised.

File: content/manager/update-deployment.md (hunk 28 above)
Comment (AviaE) [inline content/manager/update-deployment.md]: Yes. Exactly as it states.

File: content/manager/update-deployment.md (hunk 2 above)
Comment (AviaE) [inline content/manager/update-deployment.md]: Because the opening 'story' of the doc referenced 'application' in order to be most user friendly.

**Comment by AviaE** [comment]
reviewed by @nirbir

**Pull request merged**